### PR TITLE
Sass 1962 Update tax year action for employment CYA pages

### PR DESCRIPTION
**Pull request by lz1kwn**
### Description
Update tax year action for employment CYA pages
Add a link to the relevant story in Jira
[SASS-CHANGE-1962](https://jira.tools.tax.service.gov.uk/browse/SASS-1962)

##### Before Reviewing
- [x]  Have you pulled the branch down?
- [x]  Have you assigned yourself to the PR?
- [x]  Have you moved the task to “in review” on JIRA?
- [x]  Have you checked to ensure all dependencies are up to date?
- [x]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [x]  Have you run the tests?
- [x]  Have you run the journey tests?
- [x]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [x]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [x]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [x]  Have you run the journey tests? (where applicable)
- [x]  Have you addressed warnings where appropriate?
- [x]  Have you rebased against the current version of main?
- [x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [ ]  Have you checked the PR Builder passes?

### CODE REVIEW

File: app/controllers/predicates/TaxYearAction.scala [hunk 1]
```diff
@@ -38,25 +38,23 @@ class TaxYearAction @Inject()(taxYear: Int, missingTaxYearReset: Boolean)(
...
-        if (sameTaxYear || !missingTaxYearReset) {
-          Right(request)
-        } else {
-          logger.info("[TaxYearAction][refine] Tax year provided is different than that in session. Redirecting to overview.")
-          Left(
-            Redirect(appConfig.incomeTaxSubmissionOverviewUrl(taxYear))
-              .addingToSession(TAX_YEAR -> taxYear.toString)
-          )
-        }
+      //      if (!appConfig.taxYearErrorFeature || taxYear == appConfig.defaultTaxYear) {
```
Comment (georgievh) [inline app/controllers/predicates/TaxYearAction.scala:43]: I think you will need to revert this change. I think the commented sections should not be commented. Please make sure the class is same (or does the same thing) as the TaxYearAction from  personal-income-tax-submission-frontend

File: app/controllers/employment/AddEmploymentController.scala [hunk 2]
```diff
@@ -44,14 +44,14 @@ class AddEmploymentController @Inject()(implicit val cc: MessagesControllerCompo
                                         ec: ExecutionContext
                                        ) extends FrontendController(cc) with I18nSupport with SessionHelper {
 
-  def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit user =>
+  def show(taxYear: Int): Action[AnyContent] = (authorisedAction andThen TaxYearAction.taxYearAction(taxYear)).async { implicit user =>
```
Comment (georgievh) [inline app/controllers/employment/AddEmploymentController.scala:47]: I would probably create a method that would return the composed action so that we do not have to do this code here. Potential method name is authorisedTazYearAction. You can put it in a common place (an object). If you do that you will not need to inject AuthorisedAction and you can use it directly similar to the TaxYearAction. The same comment applies to all the controllers you have updated.

Further, my understanding is that the new action needs to be used in all the controllers. I think you have missed to add it to quite a few of the controllers.

**Review by georgievh — CHANGES_REQUESTED** [review]

File: app/controllers/employment/AddEmploymentController.scala (hunk 2 above)
Comment (lz1kwn) [reply app/controllers/employment/AddEmploymentController.scala]: Updated as discussed yesterday

File: app/controllers/employment/CheckEmploymentDetailsController.scala [hunk 3]
```diff
@@ -48,7 +48,7 @@ class CheckEmploymentDetailsController @Inject()(implicit val cc: MessagesContro
                                                  clock: Clock
                                                 ) extends FrontendController(cc) with I18nSupport with SessionHelper with Logging {
 
-  def show(taxYear: Int, employmentId: String): Action[AnyContent] = authAction.async { implicit user =>
+  def show(taxYear: Int, employmentId: String): Action[AnyContent] = (authAction andThen TaxYearAction.taxYearAction(taxYear)).async { implicit user =>
```
Comment (georgievh) [inline app/controllers/employment/CheckEmploymentDetailsController.scala:51]: We should probably create a composed action from the authAction and the taxYearAction and make it usable through an object. This is a common code that should be used from within all the controllers and it is good to have one early.

File: app/controllers/employment/CheckYourBenefitsController.scala [hunk 4]
```diff
@@ -51,7 +51,7 @@ class CheckYourBenefitsController @Inject()(authorisedAction: AuthorisedAction,
                                            ) extends FrontendController(mcc) with I18nSupport with SessionHelper with Logging {
 
   //scalastyle:off
-  def show(taxYear: Int, employmentId: String): Action[AnyContent] = authorisedAction.async { implicit user =>
+  def show(taxYear: Int, employmentId: String): Action[AnyContent] = (authorisedAction andThen TaxYearAction.taxYearAction(taxYear)).async { implicit user =>
```
Comment (georgievh) [inline app/controllers/employment/CheckYourBenefitsController.scala:54]: As above

File: app/controllers/expenses/CheckEmploymentExpensesController.scala [hunk 5]
```diff
@@ -49,7 +49,7 @@ class CheckEmploymentExpensesController @Inject()(authorisedAction: AuthorisedAc
                                                   implicit val mcc: MessagesControllerComponents,
                                                   implicit val ec: ExecutionContext) extends FrontendController(mcc) with I18nSupport with SessionHelper {
 
-  def show(taxYear: Int): Action[AnyContent] = authorisedAction.async { implicit user =>
+  def show(taxYear: Int): Action[AnyContent] = (authorisedAction andThen TaxYearAction.taxYearAction(taxYear)).async { implicit user =>
```
Comment (georgievh) [inline app/controllers/expenses/CheckEmploymentExpensesController.scala:52]: As above.

**Review by georgievh — CHANGES_REQUESTED** [review]
